File: backend/src/routes/meetups.js
import { Router } from 'express';
import { collections } from '../db.js';
import { body, validationResult } from 'express-validator';
import { v4 as uuid } from 'uuid';
import { verifyTokenMiddleware } from '../utils/auth.js';

function normEmail(e){ return (e||'').trim().toLowerCase(); }

const router = Router();

// Create a new one-to-one meetup (consumer<->provider)
router.post('/', verifyTokenMiddleware, [
  body('targetUserId').notEmpty().withMessage('targetUserId required'),
  body('start').notEmpty().withMessage('start ISO datetime required'),
  body('end').notEmpty().withMessage('end ISO datetime required'),
  body('title').optional().isLength({max:120}),
  body('description').optional().isLength({max:2000})
], async (req,res)=>{
  console.log('[MEETUPS][CREATE] raw body', req.body);
  const errors = validationResult(req);
  if(!errors.isEmpty()) return res.status(400).json({errors:errors.array()});

  const { targetUserId, start, end, title, description } = req.body;
  const user = req.user; // {role, id, ...}
  if(!user || !['consumer','provider'].includes(user.role)){
    return res.status(403).json({error:'only authenticated consumer or provider can create meetups'});
  }
  const startDate = new Date(start);
  const endDate = new Date(end);
  if(isNaN(startDate.getTime()) || isNaN(endDate.getTime()) || endDate <= startDate){
    return res.status(400).json({error:'invalid start/end'});
  }
  let targetUser = null;
  let targetRole = null;
  if(user.role === 'consumer'){
    targetUser = await collections().providers.findOne({ id: targetUserId });
    targetRole = 'provider';
  } else if(user.role === 'provider') {
    targetUser = await collections().consumers.findOne({ id: targetUserId });
    targetRole = 'consumer';
  }
  if(!targetUser){
    return res.status(404).json({error:'target user not found'});
  }

  const event = {
    id: uuid(),
    type: 'meetup',
    title: title || 'Meetup',
    description: description || '',
    start: startDate.toISOString(),
    end: endDate.toISOString(),
    createdAt: Date.now(),
    requesterId: user.id,
    requesterRole: user.role,
    participantId: targetUserId,
    participantRole: targetRole,
    status: 'scheduled'
  };
  await collections().events.insertOne(event);
  console.log('[MEETUPS][CREATE] created', event.id);
  res.status(201).json(event);
});

// List events for current user
router.get('/', verifyTokenMiddleware, async (req,res)=>{
  const user = req.user;
  if(!user) return res.status(401).json({error:'unauthorized'});
  const events = await collections().events.find({ $or: [ { requesterId: user.id }, { participantId: user.id } ] }, { projection: { _id:0 } }).toArray();
  res.json(events.sort((a,b)=> new Date(a.start) - new Date(b.start)));
});

// Get single event (must be participant)
router.get('/:id', verifyTokenMiddleware, async (req,res)=>{
  const user = req.user;
  const event = await collections().events.findOne({ id: req.params.id }, { projection: { _id:0 } });
  if(!event) return res.status(404).json({error:'not found'});
  if(event.requesterId !== user.id && event.participantId !== user.id && user.role !== 'admin'){
    return res.status(403).json({error:'forbidden'});
  }
  res.json(event);
});

// Update status (cancel) or details
router.patch('/:id', verifyTokenMiddleware, [
  body('status').optional().isIn(['scheduled','cancelled','completed'])
], async (req,res)=>{
  const user = req.user;
  const events = collections().events;
  const event = await events.findOne({ id: req.params.id });
  if(!event) return res.status(404).json({error:'not found'});
  if(event.requesterId !== user.id && event.participantId !== user.id && user.role !== 'admin'){
    return res.status(403).json({error:'forbidden'});
  }
  const { status, title, description, start, end } = req.body;
  if(status) event.status = status;
  if(title) event.title = title;
  if(description !== undefined) event.description = description;
  if(start){ const sd = new Date(start); if(!isNaN(sd.getTime())) event.start = sd.toISOString(); }
  if(end){ const ed = new Date(end); if(!isNaN(ed.getTime())) event.end = ed.toISOString(); }
  await events.updateOne({ id: req.params.id }, { $set: event });
  res.json(event);
});

export default router;

File: backend_python/src/routes/meetups.py
import uuid
from datetime import datetime
from typing import Optional, List
from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel, validator

from ..db import get_collections
from ..utils.auth import verify_token_middleware, normalize_email

router = APIRouter()

# Pydantic models
class MeetupCreate(BaseModel):
    targetUserId: str
    start: str  # ISO datetime string
    end: str    # ISO datetime string
    title: Optional[str] = None
    description: Optional[str] = None

    @validator('title')
    def title_length(cls, v):
        if v is not None and len(v) > 120:
            raise ValueError('Title must be 120 characters or less')
        return v

    @validator('description')
    def description_length(cls, v):
        if v is not None and len(v) > 2000:
            raise ValueError('Description must be 2000 characters or less')
        return v


class MeetupUpdate(BaseModel):
    status: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    start: Optional[str] = None
    end: Optional[str] = None

    @validator('status')
    def valid_status(cls, v):
        if v is not None and v not in ['scheduled', 'cancelled', 'completed']:
            raise ValueError('Status must be scheduled, cancelled, or completed')
        return v


def parse_datetime(date_string: str) -> datetime:
    """Parse ISO datetime string"""
    try:
        return datetime.fromisoformat(date_string.replace('Z', '+00:00'))
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid datetime format: {date_string}"
        )


@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_meetup(
    meetup_data: MeetupCreate,
    user: dict = Depends(verify_token_middleware)
):
    """Create a new one-to-one meetup (consumer<->provider)"""
    
    # Validate user role
    user_role = user.get("role")
    user_id = user.get("id")
    
    if user_role not in ["consumer", "provider"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="only authenticated consumer or provider can create meetups"
        )
    
    # Parse and validate dates
    start_date = parse_datetime(meetup_data.start)
    end_date = parse_datetime(meetup_data.end)
    
    if end_date <= start_date:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="End time must be after start time"
        )
    
    # Find target user
    collections = get_collections()
    target_user = None
    target_role = None
    
    if user_role == "consumer":
        target_user = await collections['providers'].find_one({"id": meetup_data.targetUserId})
        target_role = "provider"
    elif user_role == "provider":
        target_user = await collections['consumers'].find_one({"id": meetup_data.targetUserId})
        target_role = "consumer"
    
    if not target_user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="target user not found"
        )
    
    # Create event
    event_id = str(uuid.uuid4())
    event = {
        "id": event_id,
        "type": "meetup",
        "title": meetup_data.title or "Meetup",
        "description": meetup_data.description or "",
        "start": start_date.isoformat(),
        "end": end_date.isoformat(),
        "createdAt": datetime.now().timestamp() * 1000,  # milliseconds
        "requesterId": user_id,
        "requesterRole": user_role,
        "participantId": meetup_data.targetUserId,
        "participantRole": target_role,
        "status": "scheduled"
    }
    
    await collections['events'].insert_one(event)
    
    # Remove MongoDB _id from response
    event.pop('_id', None)
    
    return event


@router.get("/")
async def get_meetups(user: dict = Depends(verify_token_middleware)):
    """List events for current user"""
    user_id = user.get("id")
    
    if not user_id:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="unauthorized"
        )
    
    collections = get_collections()
    
    # Find events where user is requester or participant
    events = await collections['events'].find(
        {
            "$or": [
                {"requesterId": user_id},
                {"participantId": user_id}
            ]
        },
        {"_id": 0}
    ).to_list(None)
    
    # Sort by start time
    events.sort(key=lambda x: datetime.fromisoformat(x['start'].replace('Z', '+00:00')))
    
    return events


@router.get("/{event_id}")
async def get_meetup(
    event_id: str,
    user: dict = Depends(verify_token_middleware)
):
    """Get single event (must be participant)"""
    user_id = user.get("id")
    user_role = user.get("role")
    
    collections = get_collections()
    
    event = await collections['events'].find_one(
        {"id": event_id},
        {"_id": 0}
    )
    
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="not found"
        )
    
    # Check if user has access to this event
    if (event['requesterId'] != user_id and 
        event['participantId'] != user_id and 
        user_role != 'admin'):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="forbidden"
        )
    
    return event


@router.patch("/{event_id}")
async def update_meetup(
    event_id: str,
    update_data: MeetupUpdate,
    user: dict = Depends(verify_token_middleware)
):
    """Update status (cancel) or details"""
    user_id = user.get("id")
    user_role = user.get("role")
    
    collections = get_collections()
    
    # Find the event
    event = await collections['events'].find_one({"id": event_id})
    
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="not found"
        )
    
    # Check if user has permission to update this event
    if (event['requesterId'] != user_id and 
        event['participantId'] != user_id and 
        user_role != 'admin'):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="forbidden"
        )
    
    # Prepare update data
    update_fields = {}
    
    if update_data.status is not None:
        update_fields['status'] = update_data.status
    
    if update_data.title is not None:
        update_fields['title'] = update_data.title
    
    if update_data.description is not None:
        update_fields['description'] = update_data.description
    
    if update_data.start is not None:
        start_date = parse_datetime(update_data.start)
        update_fields['start'] = start_date.isoformat()
    
    if update_data.end is not None:
        end_date = parse_datetime(update_data.end)
        update_fields['end'] = end_date.isoformat()
    
    # Validate that end is still after start if both are being updated
    if 'start' in update_fields and 'end' in update_fields:
        start_dt = datetime.fromisoformat(update_fields['start'].replace('Z', '+00:00'))
        end_dt = datetime.fromisoformat(update_fields['end'].replace('Z', '+00:00'))
        if end_dt <= start_dt:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="End time must be after start time"
            )
    
    # Update the event
    if update_fields:
        await collections['events'].update_one(
            {"id": event_id},
            {"$set": update_fields}
        )
    
    # Return updated event
    updated_event = await collections['events'].find_one(
        {"id": event_id},
        {"_id": 0}
    )
    
    return updated_event
